Features/mesto update

nicholasitnikov opened the pull request.

nicholasitnikov committed:
feat: add api class

QuinnGaden committed:
feat: add-validation-class

Добавлен класс валидации форм

nicholasitnikov committed:
feat: add card class

QuinnGaden committed:
feat: Section

Добавлен класс section, реализовано добавление карточек

QuinnGaden committed:
feat: удаление карточки

Commit:
feat: add popup and popupwithform classes

QuinnGaden committed:
feat: попап редактирования профиля и аватарки

QuinnGaden committed:
feat: PopupWithImage

реализован класс PopupWithImage

nicholasitnikov committed:
feat: add userinfo class and clean code

gennady-bars commented on src/components/validate.js:
```diff
@@ -1,76 +1,76 @@
-    const inputError = form.querySelector(`${options.inputErrorClass}-${formInput.id}`);
-    inputError.textContent = formInput.validationMessage;
-    inputError.classList.remove(options.errorClass);
-}
```
`validate.js` нужно навзать `FormValidator.js`

gennady-bars commented on src/components/api.js:
```diff
@@ -1,183 +1,148 @@
+    this._apiURL = config.apiURL;
+  }
+
+  checkResponse(response) {
```
отличный метод проверки ответа

gennady-bars commented on src/components/card.js:
```diff
@@ -1,94 +1,79 @@
-        element.classList.add('place__like_active');
-    } else {
-        element.classList.remove('place__like_active');
+import { api } from './api.js';
```
с 8го спринта нельзя импортировать в файлы с классами что-либо (только класс `Popup` можно для наследования). Нужно использовать только `мягкое связывание`

gennady-bars commented on src/components/card.js:
```diff
@@ -1,94 +1,79 @@
+        this._imagePlace = this._element.querySelector('.place__image');
+
+        this._element.setAttribute('data-id', this._id);
+        this._element.querySelector('.place__image').style.backgroundImage = `url(${this._link})`;
```
` this._element.querySelector('.place__image')`

элемент картинки уже найден на 65й строчке

```js

this._imagePlace.style.backgroundImage = `url(${this._link})`;

```

gennady-bars commented on src/components/validate.js:
```diff
@@ -1,76 +1,76 @@
-    } else {
-        showErrorElement(form, formInput, options);
+    _disableSubmit() {
+        const submit = this._form.querySelector(this._submitButtonSelector);
```
`const submit = this._form.querySelector(this._submitButtonSelector);`

Все элементы, которые используются в нескольких разных методах классов или поиск которых осуществляется при каждом срабатывании метода, желательно найти 1 раз и объявить их полями класса (`классовыми` переменными с `this`). Чаще всего это удобно сделать в `конструкторе` или ` generateCard`, и тогда эти переменные будут `доступны` по всему коду класса. Это позволит сэкономить время на повторный поиск элементов, так как они 1 раз были найдены. И не надо будет их больше передавать в вызовы методов.
Эти элементы обычно `this._inputList`, `this._submitButton`, `this._form`, `this._cardImage`, `this._likeButton`

gennady-bars commented on src/components/validate.js:
```diff
@@ -1,76 +1,76 @@
-    } else {
-        enableSubmit(form, options);
+    _enableSubmit() {
+        const submit = this._form.querySelector(this._submitButtonSelector);
```
Все элементы, которые используются в нескольких разных методах классов или поиск которых осуществляется при каждом срабатывании метода, желательно найти 1 раз и объявить их полями класса (`классовыми` переменными с `this`). Чаще всего это удобно сделать в `конструкторе` или ` generateCard`, и тогда эти переменные будут `доступны` по всему коду класса. Это позволит сэкономить время на повторный поиск элементов, так как они 1 раз были найдены. И не надо будет их больше передавать в вызовы методов.
Эти элементы обычно `this._inputList`, `this._submitButton`, `this._form`, `this._cardImage`, `this._likeButton`

gennady-bars commented on src/components/validate.js:
```diff
@@ -1,76 +1,76 @@
+    }
+   
+    _setEventListeners() {
+        const formInputs = this._form.querySelectorAll(this._inputSelector);
```
Все элементы, которые используются в нескольких разных методах классов или поиск которых осуществляется при каждом срабатывании метода, желательно найти 1 раз и объявить их полями класса (`классовыми` переменными с `this`). Чаще всего это удобно сделать в `конструкторе` или ` generateCard`, и тогда эти переменные будут `доступны` по всему коду класса. Это позволит сэкономить время на повторный поиск элементов, так как они 1 раз были найдены. И не надо будет их больше передавать в вызовы методов.
Эти элементы обычно `this._inputList`, `this._submitButton`, `this._form`, `this._cardImage`, `this._likeButton`

gennady-bars commented on src/components/validate.js:
```diff
@@ -1,76 +1,76 @@
+            this._enableSubmit();
+        }
+    }
 
```
**Можно лучше**

Можно сделать отдельный  метод для очистки ошибок и управления кнопкой `resetValidation`.

В нем Вы будете пробегаться по ` this._inputList` и очищать ошибки с инпутов.

Это нужно потому, что ошибки остаются в инпутах, если пользователь начал что-то вводить в инпут неправильное, потом передумал и закрыл попап. А при повторном открытии ошибки в инпутах остались, так как они не очищаются при открытии сейчас. Вот нужно сделать это очищение и управление кнопкой (она должна быть неактивной в попапе `добавления новой карточки`).

Передавать в этот метод ничего не нужно, так как Вам нужно сделать `отдельные экземпляры` класса `FormValidator` для разных попапов с формой и эти экземпляры будут знать свои инпуты и кнопку.

Вызывать resetValidation нужно в том месте, где Вы обрабатываете нажатие на `кнопку открытия` попапа с формой.

Это будет выглядеть примерно так:

```js
 resetValidation() {
      this._toggleButtonState(); <== управляем кнопкой ==

      this._inputList.forEach((inputElement) => {
        this._hideError(inputElement) <==очищаем ошибки ==
      });

    }


```

gennady-bars commented on src/components/popup.js:
```diff
@@ -0,0 +1,27 @@
+    setEventListeners() {
+        this._removeButton = this._element.querySelector('.popup__close');
+        this._removeButton.addEventListener('click', () => this.close());
+        document.addEventListener('keydown', (e) => this._handleEscClose(e));
```
`document.addEventListener('keydown', (e) => this._handleEscClose(e));`

это нужно удалить из `setEventListeners`

По заданию и чек-листу обработчик Escape должен правильно удаляться при закрытии попапа

В таком виде невозможно удалить никакой обработчик, так как метод `bind` всегда создает `новую` ссылку на функцию, в итоге движок браузера просто не сможет найти функцию, чтобы ее удалить. Ошибки никакой браузер не выдаст, так как ему без разницы, что он не нашел функцию.

Для того чтобы ссылка стала постоянной, нужно "привязать" функцию к контексту `this` в конструкторе 1 раз с помощью `bind`.

```js
 constructor(popupSelector) {
    this._popup = …………….
    this._handleEscClose = this._handleEscClose.bind(this) <==== привязали 1 раз ===
  }


```

и теперь при открытии навешиваете и передаете только `ссылку`:


```js
 open() {
    открыли попап
    document.addEventListener('keydown', this._handleEscClose); <=== только cсылку передаете ==
    
  }

```

и удаляете при закрытии точно с такой же ссылкой:

```js
 close() {
    закрыли попап
    document.removeEventListener('keydown', this._handleEscClose); <==== такая же ссылка должна быть
  }


```

Подробнее про правильное удаление обработчиков тут: https://learn.javascript.ru/introduction-browser-events#addeventlistener

gennady-bars commented on src/components/popupWithForm.js:
```diff
@@ -0,0 +1,36 @@
+    _getInputValues() {
+        return Array.from(this._form.elements).reduce((res, current) => {
+            if(current.value === '') { return res; }
+            res[current.name] = current.value
```
`  res[current.name] = current.value`

хорошо собираете данные инпутов

gennady-bars commented on src/components/popupWithForm.js:
```diff
@@ -0,0 +1,36 @@
+        this._form = this._element.querySelector('.form');
+        this._form.addEventListener('submit', (e) => this._submitForm(e));
+        if (this._openButtonSelector) {
+            document.querySelector(this._openButtonSelector).addEventListener('click', () => this.open());
```
не нужно навешивать обработчик открытия попапов внутри `PopupWithForm`, так как при открытии еще нужно деактивировать кнопку сабмита (смотрите мой совет в `FormValidator`)

Эту обработку нужно сделать в `index.js`

gennady-bars commented on src/components/popupWithForm.js:
```diff
@@ -0,0 +1,36 @@
+        this._form.reset();
+    }
+    updateSubmitText(text) {
+        this._submitButton = this._form.querySelector('.popup__button');
```
кнопку сабмита лучше найти 1 раз в конструкторе, чтобы не тратить ресурсы на ее поиск

gennady-bars commented on src/components/popupWithImage.js:
```diff
@@ -0,0 +1,13 @@
+    super.open();
+    this._imageElement.src = src;
+    this._imageElement.alt = caption;
+    this._element.querySelector('.popup__caption').textContent = caption;
```
`'.popup__caption'` тоже можно было в конструктор передавать и найти 1 раз этот элемент там

gennady-bars commented on src/components/section.js:
```diff
@@ -0,0 +1,43 @@
+	render() {
+		this._container.innerHTML = '';
+		this._items.forEach(item => {
+			const element = this._renderer(item);
```
`const element = this._renderer(item);`

это лучше делать в `_placeInto`, тогда не нужно будет в `index.js` думать, как создавать новую карточку

gennady-bars commented on src/components/section.js:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	addItem(item) {
+		this._items.unshift(item);
```
не нужно добавлять в массив карточек новые карточки.  Нужно напрямую в `DOM` ее добавлять через `prepend`

gennady-bars commented on src/components/section.js:
```diff
@@ -0,0 +1,43 @@
+		});
+	}
+
+	renderLast() {
```
вот этот метод нужно переделать, чтобы он создавал и вставлял карточку

```js

	addItem(item) {
		const element = this._renderer(item);
		this._placeInto(element, true);	
	}

```

gennady-bars commented on src/components/section.js:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	removeItem(id) {
+		const items = document.querySelectorAll(this._itemSelector);
```
Вам не нужно находить все карточки проекта, чтобы потом найти ту, которую нужно удалить. Эту карточку можно сразу же найти прямо в `document`.

gennady-bars commented on src/components/userInfo.js:
```diff
@@ -0,0 +1,31 @@
+        this._aboutElement.textContent = this._about;
+        this._avatarElement.style.backgroundImage = `url("${this._avatar}")`;
+    }
+    setData(user) {
```
это немного лишний метод. Все нужно делать через `setUserInfo`

```js


    setUserInfo(user) {
        this._nameElement.textContent = user.name;
// и так далее

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,107 +1,148 @@
 
-enableValidation({
-    formSelector: '.form',
+const validationOptions = {
```
все константы теперь должны быть в файле `constants.js` в папке `utils`

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,107 +1,148 @@
+            removingPopup.open();            
+        },
+        handleImage: function() {
+            popupWithImage.open(this._link, this._name);
```
`this._link, this._name`

не стоит использовать `this` в обычных функциях в `index.js`. Этот контекст легко потерять. 

У Вас есть доступ к данным картинки на 42й строчке - это `card`

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,107 +1,148 @@
-    cards.forEach(card => {
-     renderPlace(createPlace(card), false);
-    })
+Promise.all([api.getInitalCards(), userInfo.getUserInfo()]).then(([cards, user]) => {
```
отлично, что получаете данные одновременно, иначе может возникнуть проблема, что `_id` пользователя еще не получили, а карточки уже пришли, и будут некорректно отображаться лайки и кнопки удаления на собственных карточках.

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,107 +1,148 @@
+        items: cards,
+        itemSelector: '.place',
+        renderer: function(card) {
+            return cardRenderer(card, user, this)
```
`this`

* не стоит использовать `this` в обычных функциях в `index.js`. Этот контекст легко потерять. 


Вы знаете, что `cardsSection` - это и есть `this`. Вот это и нужно передавать

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,107 +1,148 @@
+const removingPopup = new PopupWithForm({
+    selector: '.popup_role_remove',
+    submitHandler: async function() {
+        this.updateSubmitText('Удаление...');
```
* не стоит использовать `this` в обычных функциях в `index.js`. Этот контекст легко потерять. 

`additionPopup.updateSubmitText`

Нужно исправить все `this` в `index.js`

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,107 +1,148 @@
+            userInfo.setData(user);   
+            userInfo.setUserInfo();                    
+            this.updateSubmitText('Сохранить');
+            this.close();
```
Отлично, что закрываете попапы только при удачном ответе от сервера в блоке `then`

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,107 +1,148 @@
+            this.section.removeItem(this.card._id);
+            this.updateSubmitText('Да');
+            this.close();
+        }).catch(err => {
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented:
Здравствуйте. Работа проделана огромная, но есть некоторые недочеты:

- [x] * все файлы в папке `components` должны быть с большой буквы по чек-листу. Скрин http://joxi.ru/ZrJKJ9yuboDpgm 
- [x] * `validate.js` нужно навзать `FormValidator.js`
- [x] * с 8го спринта нельзя импортировать в файлы с классами что-либо (только класс `Popup` можно для наследования). Нужно использовать только `мягкое связывание`
- [x] * В попапе добавления новой карточки нужно `деактивировать` кнопку сабмита, если при открытии инпуты пустые, чтобы не было возможности нажать на кнопку и добавить пустую карточку без картинки и названия. А сейчас после 1го добавления карточки кнопка активируется и её нужно деактивировать опять. Тоже самое и с аватаром. Кнопки должны быть серыми (как при валидации)
- [x] * при открытии попапа профиля в инпуты нужно вставлять данные пользователя, которые отображены вверху сайта. Скрин http://joxi.ru/nAy1kG8falbxxA
- [x] * Теперь только класс `FormValidator` должен активировать / деактивировать кнопку сабмита и очищать ошибки 
- [x] * По заданию и чек-листу обработчик Escape должен правильно удаляться при закрытии попапа
- [ ] * не нужно навешивать обработчик открытия попапов внутри `PopupWithForm`, так как при открытии еще нужно деактивировать кнопку сабмита 
- [x] * не нужно добавлять в массив карточек новые карточки.  Нужно напрямую в `DOM` ее добавлять через `prepend`
- [x] * Вам не нужно находить все карточки проекта, чтобы потом найти ту, которую нужно удалить. Эту карточку можно сразу же найти прямо в `document`. 
- [x] * все константы теперь должны быть в файле `constants.js` в папке `utils`
- [x] * не стоит использовать `this` в обычных функциях в `index.js`. Этот контекст легко потерять. 
- [x] * Нужно исправить все `this` в `index.js`


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.


- [ ]

QuinnGaden committed:
fix/first-fix

исправлены имена файлов в папке components, переименован validate в FromValidator, исправлены импорты в классах, добавления карточки, удаление карточки, константы все в папке utils в файле constants.js

nicholasitnikov committed:
fix: refactor

QuinnGaden committed:
fix/contstants

Сделаны константы

gennady-bars commented on src/components/popup.js:
```diff
@@ -15,11 +18,11 @@ export default class Popup {
+        this._removeButton = this._element.querySelector(constants.popupStateSelector.popupClose);
         this._removeButton.addEventListener('click', () => this.close());
-        document.addEventListener('keydown', (e) => this._handleEscClose(e));
+        document.addEventListener('keydown', this._handleEscClose);
```
`document.addEventListener('keydown', this._handleEscClose);`

навешивать обработчик `Escape` нужно только в методе `open`, ведь в `close` его удаляете

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,56 +1,52 @@
         },
         handleImage: function() {
-            popupWithImage.open(this._link, this._name);
+            popupWithImage.open(newCard._link, newCard._name);
```
`newCard._link, newCard._name`

не нужно с нижним подчеркиванием использовать поля класса в `index.js`

Данные картинки есть в `card` на 37й строчке

```js

popupWithImage.open(card.link, card.name);

```

gennady-bars replied:
> `newCard._link, newCard._name`
> 
> не нужно с нижним подчеркиванием использовать поля класса в `index.js`
> 
> Данные картинки есть в `card` на 37й строчке
> 
> ```js
> popupWithImage.open(card.link, card.name);
> ```

Вы забыли исправить это замечание в коде

gennady-bars reviewed (commented):
Еще немного и все будет отлично:

* все файлы в папке components должны быть с большой буквы по чек-листу. Скрин http://joxi.ru/ZrJKJ9yuboDpgm
* навешивать обработчик `Escape` нужно только в методе `open`, ведь в `close` его удаляете. После первого же закрытия перестают закрываться по `Escape` попапы из-за этого
* в консоли ошибка после закрытия попапов по `Escape`. Скрин http://joxi.ru/brRxJ96TLowye2
* в консоли ошибки при любом закрытии попапа подтверждения. Скрин http://joxi.ru/MAj5QOlt1gVjD2 . Вы не указали в нем `validator` (хотя ему не нужно это, ведь у него нет инпутов и кнопку деактивировать ему нельзя). 
* Возможно, нужно удалить `validator` из класса `PopupWithForm` и вызывать `resetValidation` только при клике на кнопки открытия. Это будет самый удобный вариант

QuinnGaden committed:
fix/name-file-folder-components

QuinnGaden committed:
fix/escape-exit

nicholasitnikov committed:
Delete userInfo.js

nicholasitnikov committed:
Delete section.js

nicholasitnikov committed:
Delete popupWithImage.js

nicholasitnikov committed:
Delete popupWithForm.js

nicholasitnikov committed:
Delete popup.js

nicholasitnikov committed:
Delete card.js

nicholasitnikov committed:
Delete api.js

nicholasitnikov committed:
Update PopupWithForm.js

gennady-bars reviewed (commented):
Осталась одна мелочь:

* Кое-что забыли исправить http://joxi.ru/82QX095CVaJQom

nicholasitnikov committed:
Update index.js

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
